res/res_pjsip_device_features.c

```diff
@@ -0,0 +1,1005 @@
+		return -1;
+	}
+
+	AST_STANDARD_APP_ARGS(args, data);
+	*buf = '\0';
+
+	if (ast_strlen_zero(args.endpoint) || ast_strlen_zero(args.feature)) {
+		ast_log(LOG_WARNING, "An endpoint and feature are required\n");
+		return -1;
+	}
+
+	/* We don't actually care if the subscription currently exists or not.
+	 * Just try to see if there's anything cached at the moment. */
+
+	/* Retrieve what the user wants from the cache. */
+	if (!strcasecmp(args.feature, "donotdisturb")) {
+		dbfamily = "subscription_device_features/donotdisturb";
+	} else if (!strcasecmp(args.feature, "callforwardingalways")) {
+		dbfamily = "subscription_device_features/callforwardingalways";
+	} else if (!strcasecmp(args.feature, "callforwardingbusy")) {
+		dbfamily = "subscription_device_features/callforwardingbusy";
+	} else if (!strcasecmp(args.feature, "callforwardingnoanswer")) {
+		dbfamily = "subscription_device_features/callforwardingnoanswer";
+	} else if (!strcasecmp(args.feature, "ringcount")) {
+		dbfamily = "subscription_device_features/ringcount";
+	} else {
+		ast_log(LOG_WARNING, "Invalid feature: %s\n", args.feature);
+		res = -1;
+	}
+
+	if (!res && ast_db_get(dbfamily, args.endpoint, buf, len - 1)) {
+		ast_debug(1, "No cached data available for endpoint %s, feature %s\n", args.endpoint, args.feature);
+		res = -1;
+	}
+
+	return res;
+}
+
+static int func_features_write(struct ast_channel *chan, const char *function, char *data, const char *value)
+{
+	struct feature_state_subscription *feature_state_sub;
+	int res = 0;
+	char *tmp;
+
+	AST_DECLARE_APP_ARGS(args,
+		AST_APP_ARG(endpoint);
+		AST_APP_ARG(feature);
+	);
+	AST_DECLARE_APP_ARGS(args2,
+		AST_APP_ARG(status);
+		AST_APP_ARG(rings);
+	);
+
+	if (ast_strlen_zero(data)) {
+		ast_log(LOG_WARNING, "Missing arguments\n");
+		return -1;
+	}
+
+	AST_STANDARD_APP_ARGS(args, data);
+	tmp = ast_strdupa(S_OR(value, ""));
+	AST_STANDARD_APP_ARGS(args2, tmp);
+
+	if (ast_strlen_zero(args.endpoint) || ast_strlen_zero(args.feature)) {
+		ast_log(LOG_WARNING, "An endpoint and feature are required\n");
+		return -1;
+	}
+
+	if (ast_strlen_zero(args2.status)) {
+		args2.status = "";
+	}
+
+	feature_state_sub = feature_state_sub_by_endpoint(args.endpoint);
+	if (!feature_state_sub) {
+		ast_log(LOG_WARNING, "No device feature subscription for %s\n", args.endpoint);
+		return -1;
+	}
+	/* Terminated subscriptions are no longer associated with a valid tree. Do not queue notify_task. */
+	if (ast_sip_subscription_is_terminated(feature_state_sub->sip_sub)) {
+		ast_log(LOG_WARNING, "Subscription for %s is already terminated\n", args.endpoint);
+		ao2_ref(feature_state_sub, -1);
+		return -1;
+	}
+
+	/* Get the new status, cache in AstDB, and then send a NOTIFY. */
+	if (!strcasecmp(args.feature, "donotdisturb")) {
+		int new_dnd_status = ast_true(args2.status) ? 1 : 0; /* Use 1, not -1 */
+		DEVICE_FEATURE_UPDATE(args.endpoint, "donotdisturb", update_needed_dnd, dnd, new_dnd_status, new_dnd_status ? "1" : "0");
```

Comment:
There is no protection or serialization around these. Is it possible for updates to occur at the same time?